### DISCUSSION TITLE
When compiler OW-V2 can it generate 64 bit executables on Windows ?

### DISCUSSION THREAD

**Discussion by tdechaize** [open-watcom/open-watcom-v2#1288] [Ideas]
Hi,
Today, OW is very limited to generate only 32 bit executable (.ie. format PE) on Windows (and also on Linux ..., format ELF 32 bit).
Is it investigate to support PE+ format of executables (64 bit) on Windows (and also format ELF+ 64 bit on Linux) ?
For me, this will may very succesfull update to preserve future of this compiler.
Any opinions ?
Sincerly.

**Comment by jmalak** [comment]
some necessary steps were done:
- porting OW code to 64-bit that now we have 32-bit and 64-bit version executable
- porting OW tools to support 64-bit objects, it is partialy done for librarian, linker and resource compiler
- create new code generator for 64-bit Intel architecture, not yet started

**Reply by tdechaize** [reply]
Hi,

Thank's.

Three steps :

    - first, it's realized, I think, because two different directory of binary (binnt and binnt64 for windows version of OW)

   - second, partialy OK.  How many tools don't meet 64 bit version ?

   - third, if you are interested a library exist : https://www.codeproject.com/Articles/5205732/libpe-PE32-PE32plus-Binaries-Parsing-Library
   
Sincerly.

Thierry D.

**Comment by st-hansdraugen** [comment]
This is lower levels than I normally engage in, because I do not master them. But from memory I have a notion that there is a PE executable variety that is considered "64"-bits in WIN64, but needs only the pointers to be changed to 64-bits. So a new CG needs that modification only, to gap a present need until a full 64-bits CG can be written. This may not be what the original poster wants, and I have no idea how much work it represents, but might be a possibility for an incremental approach.

**Reply by jmalak** [reply]
It is not about output executable format for 64-bit PE, which is supported by OW tools now (by example resource compiler).
The main task is to create Code Generator which generate 64-bit instructions in 64-bit segments instead of 32-bit instruction in 32-bit segments. It means addressing mode must be 64-bit.

**Comment by st-hansdraugen** [comment]
Thank you for the clarification. It is more complex than I have lead myself to believe. Do I understand it correctly (in a broad sense) that what is needed is a CG generating x86-64-v1 based CPU instructions and correspondingly calling 64-bit based OS APIs?

**Reply by jmalak** [reply]
Yes, it requires Code Generator for x64 instruction and next build C run-time library for 64-bit targets. 
CRTL will require some changes due to assembly code use. 
But most of code is in C that minimal changes will be required, only when code is bit size dependant.
